File: Lecture9/ridders.cpp
// This file basic.hpp provides functions for basic numerical algorithms

#include <algorithm>
#include <cmath>
#include <complex>
#include <cstdlib>
#include <ctime>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <limits>
#include <sstream>
#include <string>
#include <vector>
using namespace std;

double diff_Ridders(
    double (*f)(double),        // name of function to be differentiated
    double x,                   // input: the point at which df/dx is required
    double h,                   // input: suggestion for an initial step size
    double& error)              // output: estimate of error by algorithm
{
    if (h == 0.0) {
        cerr << "diff_Ridders: h must be non-zero\n";
        exit(1);
    }
    const int n = 10;           // dimension of extrapolation table
    double a[n][n];             // extrapolation table
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            a[i][j] = 0;
    a[0][0] = (f(x + h) - f(x - h)) / (2 * h);
    double answer = 0;
    error = numeric_limits<double>::max();
    for (int i = 0; i < n; i++) {
        h /= 1.4;
        a[0][i] = (f(x + h) - f(x - h)) / (2 * h);
        double fac = 1.4 * 1.4;
        for (int j = 1; j <= i; j++) {
            a[j][i]=(a[j-1][i] * fac - a[j-1][i-1]) / (fac - 1);
            fac *= 1.4 * 1.4;
            double err = max(abs(a[j][i] - a[j-1][i]),
                             abs(a[j][i] - a[j-1][i-1]));
            if (err <= error) {
                error = err;
                answer = a[j][i];
            }
        }
        if (abs(a[i][i] - a[i-1][i-1]) >= 2 * error)
            break;
    }
    return answer;
}

double test_ridders_x2( double x ) { return x*x; }


int main(int argc, char ** argv) 
{
  cout <<  "Testing Ridder's algorithm for computing numeric derivatives" << endl;

  double error = 0.0;
  double h = 0.01;

  char buff[1000];
  for ( double x = 2.0; x <= 5.0 ; x+= 1.0) {
    double xprime = diff_Ridders( std::exp, x, h, error );
    sprintf( buff, "%2.0f : True value = %38.32f, Ridder = %38.32f, exp error = %38.32f, obs error = %38.32f", x, std::exp(x), xprime, h*h*h*h, error );
    cout << buff << endl;

    xprime = diff_Ridders( std::sin, x, h, error );
    sprintf( buff, "%2.0f : True value = %38.32f, Ridder = %38.32f, sin error = %38.32f, obs error = %38.32f", x, std::cos(x), xprime, h*h*h*h, error );
    cout << buff << endl;


    xprime = diff_Ridders( test_ridders_x2, x, h, error );
    sprintf( buff, "%2.0f : True value = %38.32f, Ridder = %38.32f, sin error = %38.32f, obs error = %38.32f", x, 2*x, xprime, h*h*h*h, error );
    cout << buff << endl;

  }

  return 0;

}

File: Lecture9/ridders.py
from math import sqrt, sin, cos, exp
from numpy import array, nan_to_num, inf
 
def ridder(f,x,h):
    ''' f     : name of function to be differentiated
        x     : the point at which df/dx is required
        h     : suggestion for an initial step size
        error : estimate of error by algorithm
    '''

    if h == 0.0 :
        print "diff_Ridders: h must be non-zero"
        exit

    n = 10           # dimension of extrapolation table
    a = array(  [[0.0] * n] *n )             # extrapolation table

    a[0][0] = (f(x + h) - f(x - h)) / (2 * h)
    answer = 0.0
    error = nan_to_num( inf ) / 2.0  # get a large value for the error
    for i in xrange(n) : 
        h /= 1.4
        a[0][i] = (f(x + h) - f(x - h)) / (2 * h)
        fac = 1.4 * 1.4
        for j in range(1,i+1) :
            a[j][i]=(a[j-1][i] * fac - a[j-1][i-1]) / (fac - 1)
            fac *= 1.4 * 1.4
            err = max(abs(a[j][i] - a[j-1][i]),
                      abs(a[j][i] - a[j-1][i-1]))
            if err <= error :
                error = err
                answer = a[j][i]

        if abs(a[i][i] - a[i-1][i-1]) >= 2 * error :
            break
    return answer,error


def test_ridders_x2 ( x ) :
    return x*x

print "Testing Ridder's algorithm for computing numeric derivatives"

error = 0.0
h = 0.01

for x in [2.0,3.0,4.0,5.0] :
    xprime,error =  ridder( exp, x, h)
    print 'exp({0:1.0f}) : True value = {1:38.32f}, Ridder = {2:38.32f}, exp error = {3:38.32f}, obs error = {4:38.32f}'.format( x, exp(x), xprime, h**4, error)
    xprime,error =  ridder( sin, x, h)
    print 'sin({0:1.0f}) : True value = {1:38.32f}, Ridder = {2:38.32f}, exp error = {3:38.32f}, obs error = {4:38.32f}'.format( x, cos(x), xprime, h**4, error)
    xprime,error =  ridder( test_ridders_x2, x, h)
    print 'x*2({0:1.0f}) : True value = {1:38.32f}, Ridder = {2:38.32f}, exp error = {3:38.32f}, obs error = {4:38.32f}'.format( x, 2*x, xprime, h**4, error)
    
